testing/base/source_gen.cpp

```diff
@@ -0,0 +1,676 @@
+  int param_lines = 0;
+  for (int num_params : llvm::seq_inclusive(0, params.max_params)) {
+    if (num_params > NumSingleLineFunctionParams) {
+      param_lines += (num_params + MaxParamsPerLine - 1) / MaxParamsPerLine;
+    }
+  }
+  return 1.0 + static_cast<double>(param_lines) / (params.max_params + 1);
+}
+
+static auto EstimateAvgMethodDeclLines(SourceGen::MethodDeclParams params)
+    -> double {
+  // Currently model a uniform distribution [0, max] parameters. Assume a line
+  // break before the first parameter for >2 and after every 4th.
+  int param_lines = 0;
+  for (int num_params : llvm::seq_inclusive(0, params.max_params)) {
+    if (num_params > NumSingleLineMethodParams) {
+      param_lines += (num_params + MaxParamsPerLine - 1) / MaxParamsPerLine;
+    }
+  }
+  return 1.0 + static_cast<double>(param_lines) / (params.max_params + 1);
+}
+
+// Note that this should match the heuristics used when formatting.
+// TODO: See top-level TODO about line estimates and formatting.
+static auto EstimateAvgClassLines(SourceGen::ClassParams params) -> double {
+  // Blank line, comment line, and class open line.
+  double avg = 3.0;
+
+  // One comment line and blank line per function, plus the function lines.
+  avg +=
+      (2.0 + EstimateAvgFunctionDeclLines(params.public_function_decl_params)) *
+      params.public_function_decls;
+  avg += (2.0 + EstimateAvgMethodDeclLines(params.public_method_decl_params)) *
+         params.public_method_decls;
+  avg += (2.0 +
+          EstimateAvgFunctionDeclLines(params.private_function_decl_params)) *
+         params.private_function_decls;
+  avg += (2.0 + EstimateAvgMethodDeclLines(params.private_method_decl_params)) *
+         params.private_method_decls;
+
+  // A blank line and all the fields (if any).
+  if (params.private_field_decls > 0) {
+    avg += 1.0 + params.private_field_decls;
+  }
+
+  // No need to account for the class close line, we have an extra blank line
+  // count for the last of the above.
+  return avg;
+}
+
+auto SourceGen::GenAPIFileDenseDecls(int target_lines, DenseDeclParams params)
+    -> std::string {
+  std::string source;
+  llvm::raw_string_ostream os(source);
+
+  double avg_class_lines = EstimateAvgClassLines(params.class_params);
+  int num_classes = static_cast<double>(target_lines) / avg_class_lines;
+  int expected_lines = num_classes * avg_class_lines;
+  os << "// Generated " << (!IsCpp() ? "Carbon" : "C++") << " source file.\n";
+  os << llvm::formatv("// {0} target lines: {1} classes, {2} expected lines",
+                      target_lines, num_classes, expected_lines)
+     << "\n";
+  os << "//\n// Generating as an API file with dense declarations.\n\n";
+
+  auto class_gen_state = GetClassGenState(num_classes, params.class_params);
+  llvm::ListSeparator line_sep("\n");
+  for ([[maybe_unused]] int i : llvm::seq(num_classes)) {
+    os << line_sep;
+    GenerateClassDef(params.class_params, class_gen_state, os);
+  }
```

Comment:
```suggestion
  os << "//\n// Generating as an API file with dense declarations.\n";

  auto class_gen_state = GetClassGenState(num_classes, params.class_params);
  for ([[maybe_unused]] int i : llvm::seq(num_classes)) {
    os << '\n';
    GenerateClassDef(params.class_params, class_gen_state, os);
  }
```
Optional, but I find this a bit simpler. (It also better handles `num_classes == 0` but I think we probably don't care.)

Reply:
Did something roughly similar as part of getting the estimation correct.

I use a string literal rather than a character literal though as I find that a bit clearer when streaming.